### PR TITLE
Rework Match.group handling

### LINKED ISSUE #3902: Match.group can return None
All usages of `Match.group` appear to return strings however, for optional named groups it can return None:

```python
import re
assert re.match("(?P<numbers>[1-9]+)?", "abc").group("numbers") is None
```

### PR DESCRIPTION
**Pull request by srittau**
Standardize `group()`, `groups()`, `groupdict()`, and `__getattr__()` to return `AnyStr | Any` instead of `AnyStr`. Also special case `group(0)` and `__getattr__(0)` to always return `AnyStr`. Correctly handle defaults. Use PEP 604 for unions in changed lines.

*Edited* to include changes after review below.

Fixes #3902

### CODE REVIEW

File: stdlib/typing.pyi [hunk 1]
```diff
@@ -2,6 +2,7 @@ import collections  # Needed by aliases like DefaultDict, see mypy issue 2986
 import sys
 from abc import ABCMeta, abstractmethod
 from types import BuiltinFunctionType, CodeType, FrameType, FunctionType, MethodType, ModuleType, TracebackType
+from typing_extensions import Literal as _Literal
```
Comment (Akuli) [inline stdlib/typing.pyi:5]: Why is this necessary?

**Comment by Akuli** [comment]
I'm arfaid this will cause dead code warnings in mypy, if you do something like this:

```
foo = match.group(1)
if foo is None:
    ...
else:
    ...
```

Having type checkers say "this is a string" has caused confusion before. See https://github.com/python/typeshed/pull/5269#issuecomment-830225377 for example.

File: stdlib/typing.pyi (hunk 1 above)
Comment (srittau) [reply stdlib/typing.pyi:5]: Because Python 3.6 and 3.7 doesn't have `Literal`.

Comment (Akuli) [reply stdlib/typing.pyi:5]: I meant, why do we need to import it `as _Literal`.

**Comment by srittau** [comment]
Maybe we should just use `AnyStr | Any` here as well. I will try that.

**Comment by Akuli** [comment]
Seems like it would also be better for mypyc (it's just one tool that uses typeshed, but an important tool): https://github.com/python/typeshed/pull/5269#issuecomment-830236965

File: stdlib/typing.pyi (hunk 1 above)
Comment (srittau) [reply stdlib/typing.pyi:5]: Because `typing` defines Literal itself and this would be a conflict. The alternative is to use a conditional import on < Python 3.8.

Comment (Akuli) [reply stdlib/typing.pyi:5]: Nevermind. Somehow I thought this was `re.pyi`.

File: stdlib/typing.pyi [hunk 2]
```diff
@@ -563,19 +564,29 @@ class Match(Generic[AnyStr]):
...
     @overload
-    def group(self, __group1: Union[str, int], __group2: Union[str, int], *groups: Union[str, int]) -> Tuple[AnyStr, ...]: ...
-    def groups(self, default: AnyStr = ...) -> Sequence[AnyStr]: ...
-    def groupdict(self, default: AnyStr = ...) -> dict[str, AnyStr]: ...
+    def group(self, __group: str | int) -> AnyStr | Any: ...
+    @overload
+    def group(self, __group1: str | int, __group2: str | int, *groups: str | int) -> Tuple[AnyStr | Any, ...]: ...
+    # Each item of groups()'s return tuple is either "AnyStr" or
+    # "AnyStr | None", depending on the pattern.
+    def groups(self, default: AnyStr = ...) -> Tuple[AnyStr | Any, ...]: ...
```
Comment (Akuli) [inline stdlib/typing.pyi:576]: The `default` annotation is wrong:
- If `default` is None (the default), then the resulting tuple can contain `None`s.
- If `default` is `AnyStr`, then the resulting tuple will not contain `None`s.

```
>>> print(re.fullmatch('a(b)?', 'a').groups())
(None,)
>>> print(re.fullmatch('a(b)?', 'a').groups(default=None))
(None,)
>>> print(re.fullmatch('a(b)?', 'a').groups(default="foo"))
('foo',)
>>> 
```

The default can in fact be any object, so maybe this with a TypeVar named `_T`?

```suggestion
    @overload
    def groups(self, default: None = ...) -> Tuple[AnyStr | Any, ...]: ...
    @overload
    def groups(self, default: _T) -> Tuple[AnyStr | _T, ...]: ...
```

This also applies to `.groupdict()`.

Comment (srittau) [reply stdlib/typing.pyi]: I actually changed the first overload not to accept any argument, with the rationale that if someone uses `m.groups(default=None)`, they are prepared to deal with `None` values in the result tuple.

**Comment by srittau** [comment]
The primer output looks tolerable in my opinion, but we should wait to merge this to after the next mypy release, though, to minimize disruption.

**Comment by srittau** [comment]
Now that mypy has released, I'd consider this ready for merge.
